File: ARIMAestim.m
function m = ARIMAestim(m)
% m = ARIMAestim(m)
%
% ARIMAestim Estimates and forecasts a time series using an ARIMA model
%
% m is an object containing the fields. See help for ARIMAmodel.
%
    if isempty(m.u)
        u = m.u;
    else
        if isvector(m.u)
            u = reshape(m.u, 1, length(m.u));
        else
            [nu1, nu2] = size(m.u);
            u = double(m.u);
            u = reshape(u, nu1, nu2);
        end
    end
    % Call ARIMAc function (assumed to be implemented elsewhere)
    [m.p, yFor, yForV, ySimul, m.lambda, m.model, m.cnst, m.u, m.BIC, ...
        m.AIC, m.AICc, m.IC, m.table, m.v] = ...
             ARIMAc('estimate', double(m.y), u, m.model, m.cnst, m.s, ...
                    m.criterion, m.h, m.verbose, m.lambda, m.maxOrders, ...
                    m.bootstrap, m.nSimul, m.fast, m.identDiff, m.identMethod);
    if isempty(m.p)
        error('Error in ARIMA estimation!!');
    else
        lu = size(m.u, 2); % Ensure the correct dimension is used
        if lu > 0
            m.h = lu - length(m.y);
        end
        if istimetable(m.y) && m.h > 0
            fake = [m.y; NaN]; % Create a fake series with an extra NA
            m.yFor = array2timetable(yFor, 'RowTimes', fake.Properties.RowTimes);
            m.yForV = array2timetable(yForV, 'RowTimes', fake.Properties.RowTimes);
            if m.bootstrap
                m.ySimul = array2timetable(ySimul, 'RowTimes', fake.Properties.RowTimes);
            end
        elseif m.h > 0
            m.yFor = yFor;
            m.yForV = yForV;
            m.ySimul = ySimul;
        end
        return;
    end
end

File: R/ARIMAestim.R
#' @title ARIMAestim
#' @description Estimates and forecasts ARIMA models
#'
#' @details \code{ARIMAestim} estimates and forecasts a time series using 
#' an ARIMA model
#'
#' @param m an object of type \code{ARIMA} created with \code{ARIMAforecast}
#' 
#' @return The same input object with the appropriate fields 
#' filled in, in particular:
#' \item{p}{Estimated parameters}
#' \item{yFor}{Forecasted values of output}
#' \item{yForV}{Variance of forecasted values of output}
#' \item{ySimul}{Bootstrap simulations for forecasting distribution evaluation}
#' 
#' @export
#' 
ARIMAestim = function(m){
    if (is.null(m$u))
        u = m$u
    else {
        if (is.vector(m$u)){
            u = matrix(m$u, 1, length(m$u))
        } else {
            nu = dim(m$u)
            u = as.numeric(m$u);
            u = matrix(u, nu[1], nu[2])
        }
    }
    output = ARIMAc("estimate", as.numeric(m$y), u, m$model, m$cnst, m$s, m$criterion,
                    m$h, m$verbose, m$lambda, m$maxOrders, m$bootstrap, m$nSimul,
                    m$fast, m$identDiff, m$identMethod)
    if (length(output) == 1){   # ERROR!!
            stop()
    } else {
            m$p = output$p
            m$ySimul = output$ySimul
            m$lambda = output$lambda
            m$model = output$orders
            m$cnst = output$cnst
            m$u = output$u
            m$BIC = output$BIC
            m$AIC = output$AIC
            m$AICc = output$AICc
            m$IC = output$IC
            # m$betaAug = output$betaAug
            # m$betaAugVar = output$betaAugVar
            if (!is.null(dim(m$u))){
                    lu = dim(m$u)[1]
                    if (lu == 1)
                            lu = dim(m$u)[2]
            } else {
                    lu = length(m$u)
            }
            if (lu > 0)
                m$h = lu - length(m$y)
            if (is.ts(m$y) && m$h > 0){
                fake = ts(c(m$y, NA), start = start(m$y), frequency = frequency(m$y))
                m$yFor = ts(output$yFor, start = end(fake), frequency = frequency(m$y))
                m$yForV = ts(output$yForV, start = end(fake), frequency = frequency(m$y))
                if (m$bootstrap)
                    m$ySimul = ts(output$ySimul, start = end(fake), frequency = frequency(m$y))
            } else if (m$h > 0) {
                m$yFor = output$yFor
                m$yForV = output$yForV
                m$ySimul = output$ySimul
            }
            return(m)
    }
}